File: src/main/java/com/cqjtu/cssl/service/ProjectService.java
package com.cqjtu.cssl.service;

import com.cqjtu.cssl.entity.Project;

import java.util.List;

/**
 * 项目卡片信息服务接口
 *
 * @author suwen
 * @date 2020/2/6 3:57 下午
 */
public interface ProjectService {

  /**
   * 新增一条项目卡片数据
   *
   * @author suwen
   * @date 2020/2/6 3:57 下午
   * @param project 项目卡片
   */
  void addProject(Project project);

  /**
   * 删除一条项目卡片数据记录
   *
   * @author suwen
   * @date 2020/2/6 3:57 下午
   * @param proId 项目卡片编号
   */
  void removeProject(Integer proId);

  /**
   * 更新一条项目卡片记录
   *
   * @author suwen
   * @date 2020/2/6 3:57 下午
   * @param proId 项目卡片编号
   */
  void updateProject(Project proId);

  /**
   * 获取一条项目卡片记录
   *
   * @author suwen
   * @date 2020/2/6 3:57 下午
   * @param proId 项目卡片编号
   * @return Project
   */
  Project getProjectById(Integer proId);

  /**
   * 获取所有项目卡片记录
   *
   * @author suwen
   * @date 2020/2/6 3:57 下午
   * @return List<com.cqjtu.cssl.entity.Project>
   */
  List<Project> loadAll();
}

File: src/main/webapp/LabAdmin2.0/src/app/service/project.service.ts
import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {HttpClient} from '@angular/common/http';
import {Exp, ProjectItem} from '../enity/project';
import {environment} from '../../environments/environment';
import {NzMessageService} from 'ng-zorro-antd';
import {HandleError} from './handle-error';
import {catchError, tap} from 'rxjs/operators';
import {result} from "../enity/result";

@Injectable({
  providedIn: 'root'
})
export class ProjectService extends HandleError {

  constructor(
    private http: HttpClient, message: NzMessageService) {
    super(message);
  }

  private PROJECT_API = `${environment.apiUrl}/project`;
  private ITEM_API = `${environment.apiUrl}/projectItem`;

  /**
   * @description 增加项目卡片
   *
   * @param project 项目卡片信息
   * @return 执行结果（true: 成功；false: 失败）
   * @author suwen
   * @date 2020/5/27 上午10:49
   */
  addProject(project: Exp): Observable<result> {
    return this.http.post<result>(this.PROJECT_API + '/addProject', project).pipe(
      tap(response => {
        if (response.success) {
          this.success(response.message);
        } else {
          this.error('增加项目卡片失败');
        }
      }),
      catchError(this.handleError<result>('增加项目卡片'))
    );
  }

  /**
   * @description 增加实验卡片项目项
   *
   * @param projectItems 实验卡片项目项
   * @return 执行结果（true: 成功；false: 失败）
   * @author suwen
   * @date 2020/5/27 上午10:51
   */
  addProjectItems(projectItems: ProjectItem[]): Observable<result> {
    return this.http.post<result>(this.ITEM_API + '/addProjectItems', projectItems).pipe(
      tap(response => {
        if (response.success) {
          this.success(response.message);
        } else {
          this.error('增加实验卡片项目项失败');
        }
      }),
      catchError(this.handleError<result>('增加实验卡片项目项'))
    );
  }

  /**
   * @description 根据教师编号获取项目卡片
   *
   * @param tid 教师编号
   * @param term 学期
   * @return 项目卡片
   * @author suwen
   * @date 2020/5/27 上午10:53
   */
  getProjects(tid: string, term: string): Observable<result> {
    const url = `${this.PROJECT_API}/getProject`;


    return this.http.get<result>(url, {params: {tid, term}}).pipe(
      tap(response => {
        if (response.success) {
          this.success(response.message);
        } else {
          this.error(`根据教师编号获取项目卡片失败，教师编号：${tid}`);
        }
      }),
      catchError(this.handleError<result>(`根据教师编号获取项目卡片失败,教师编号${tid}`))
    );
  }

  /**
   * @description 根据教师编号获取项目卡片项
   *
   * @param proId 项目卡片编号
   * @return 项目卡片项
   * @author suwen
   * @date 2020/5/27 上午10:55
   */
  getProjectItems(proId: number): Observable<result> {
    const url = `${this.ITEM_API}/getProjectItem/${proId}`;
    return this.http.get<result>(url).pipe(
      tap(response => {
        if (response.success) {
          this.success(response.message);
        } else {
          this.error(`根据教师编号获取项目卡片项，项目卡片编号：${proId}`);
        }
      }),
      catchError(this.handleError<result>(`根据教师编号获取项目卡片项，项目卡片编号：${proId}`))
    );
  }

}
